Tests/tinys3Tests/requests/ListMultipartUploadsRequestTests.swift

```diff
@@ -0,0 +1,66 @@
+import XCTest
+@testable import tinys3
+
+// swiftlint:disable line_length
+final class ListMultipartUploadsRequestTests: XCTestCase, RequestTest {
+    let request = AWSRequest.listMultipartUploadsRequest(
+        bucket: "examplebucket",
+        key: "images/xcode-14.3.vmtemplate.aar",
+        credentials: .testDefault,
+        date: .testDefault
+    )
+
+    func testThatCanonicalUriIsCorrect() throws {
+        XCTAssertEqual("/", request.canonicalUri)
+    }
+
+    func testThatCanonicalQueryStringIsCorrect() throws {
+        XCTAssertEqual("prefix=images%2Fxcode-14.3.vmtemplate.aar&uploads=", request.canonicalQueryString)
+    }
+
+    func testThatCanonicalHeaderStringIsCorrect() throws {
+        XCTAssertEqual("""
+host:examplebucket.s3.amazonaws.com
+x-amz-content-sha256:e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
+x-amz-date:20130524T000000Z
+""", request.canonicalHeaderString)
+    }
+
+    func testThatSignedHeaderStringIsCorrect() throws {
+        XCTAssertEqual("host;x-amz-content-sha256;x-amz-date", request.signedHeaderString)
+    }
+
+    func testThatCanonicalRequestIsValid() throws {
+        XCTAssertEqual("""
+GET
+/
+prefix=images%2Fxcode-14.3.vmtemplate.aar&uploads=
+host:examplebucket.s3.amazonaws.com
+x-amz-content-sha256:e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
+x-amz-date:20130524T000000Z
+
+host;x-amz-content-sha256;x-amz-date
+e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
+""", request.canonicalRequest)
+    }
+
+    func testThatStringToSignIsValid() throws {
+        XCTAssertEqual("""
+AWS4-HMAC-SHA256
+20130524T000000Z
+20130524/us-east-1/s3/aws4_request
+90f9822ca97b1ca0de0d1b0b4afb8137cf97ad4d07396aa6686a1cfacb6dc124
+""", request.stringToSign)
+    }
+
+    func testThatSignatureIsValid() throws {
+        XCTAssertEqual("5feb3c99fc520055ad41dcb6c99dfb714d6bfdf670eb44a4dd14b289b9a5b4ed", request.signature)
```

Comment:
Is the expected signature text here generated using the `stringToSign` value in the above test and the `request.credentials`?

Is there a way to get the expected signature without running any hostmgr code? Because if for some reason, the `stringToSign` or `credentials` is changed, we'll need to update this signature here, and I guess we should just run the "testee" code and use its result.

Also, consiering the correlation between these two tests, maybe we can merge them into one?

Reply:
> Is the expected signature text here generated using the stringToSign value in the above test and the request.credentials?

Yes. In practice `AWSRequest.signature` simply call's its `AWSRequestSigner`'s `.sign(string: stringToSign)`, so this kinda tests the `AWSRequestSigner` more than the `AWSRequest` itself, though it also confirms that the request uses the expected signer.

> Is there a way to get the expected signature without running any hostmgr code? Because if for some reason, the stringToSign or credentials is changed, we'll need to update this signature here, and I guess we should just run the "testee" code and use its result.

Yeah, this is tricky. The only way to ensure this test works is to run it, and if it fails, copy the actual result of the signer as the expected value for the test… so in a way we're writing a test to fit the result instead of the other way around.

There's no real way to get the expected signature without running any `hostmgr` code, as the only way to confirm or deny if our implementation that computes the signature is correct or not is by actually sending a signed request to AWS and see if their server accepts or rejects it. (Btw, this is how I found out that our implementation was incorrect back in https://github.com/Automattic/hostmgr/pull/89#discussion_r1533114140)

While I'm not loving this need to run the test to find out and adjust the expected value—or to send the request to AWS live and check if it rejects it—I'm not sure there's any better way to write a test against the signature algorithm implementation.

> Also, consiering the correlation between these two tests, maybe we can merge them into one?

That could be an option, but this also tests that the `AWSRequest.signature` implementation indeed calls the expected `AWSRequestSigner` implementation (the one implementing AWS4's HMAC algorithm), and that signer uses not only the proper `stringToSign` but also the proper request date and credentials to compute its value. So I think this still has value to keep this as a separate test method?

Besides, all those requests-related tests (tinys3Tests/requests/*Tests.swift`) conform to our `RequestTest` protocol which expects those test methods 🙃

Reply:
I tried to use `aws s3 presign` to generate a signed URL, which is essentially what we are doing here. However, the cli doesn't accept an `AWSRequest.date` equivalent (it probably always uses the "now" timestamp).

```
AWS_ACCESS_KEY_ID=AKIAIOSFODNN7EXAMPLE AWS_SECRET_ACCESS_KEY=wJalrXUtnFEMI/K7MDENG/bPxRfiCYEXAMPLEKEY AWS_DEFAULT_REGION=us-east-1 aws s3 presign s3://examplebucket/images/xcode-14.3.vmtemplate.aar
```

There are AWS SDKs in many different languages, that must have implemented signing requests, where there are more granular control over the parameters. We could potentially add some comment here with ruby/python code that calls AWS SDK with the parameters in `AWSRequest` to get an expected signature.

Reply:
`aws s3 presign` is a bit different, in that it implements [the "query string authentication alternative"](https://docs.aws.amazon.com/AmazonS3/latest/userguide/RESTAuthentication.html#RESTAuthenticationQueryStringAuth) approach, as opposed to create a signature for an arbitrary request and provide that signature via `Authorization` header. This also only works for GET requests to access an S3 object—as in [`PresignedURLGetRequestTests.swift`](https://github.com/Automattic/hostmgr/blob/trunk/Tests/tinys3Tests/presign/PresignedURLGetRequestTests.swift)—as it's designed to create URLs that can be shared as direct links (e.g. over Slack or as an HTML link in web page) to S3 objects.

---

So instead, in 555dd80f49b9798d6e3f0be84959416c3a8d5ec7, I've added a `printDebugSigningRubyCode()` method on `AWSRequest` in `TestHelpers.swift` that prints Ruby code to be pasted in your favorite Ruby interpreter (`irb`, `pry`, …) and will print you the expected values for those test methods, as per returned by the official `aws-sigv4` Ruby gem.

So whenever you want to implement a new test, like I did in 7184f82a36dc8bf3728924924b73bd6a8678ca42, you can put a breakpoint in one of the test methods, call `po printDebugSigningRubyCode()`, copy/paste the code from the Console to `irb` to run it, and then copy/paste the answers printed from Ruby back to your Swift code 🎉

Reply:
That looks great! Thanks for looking into it 👍